peps/pep-0740.rst

```diff
@@ -205,40 +177,224 @@ changes to it:
-* Each attestation value **MUST** be verifiable by the index. If the index fails
+* The new ``attestations`` field **SHALL** be a JSON array.
+* The ``attestations`` array **SHALL** have one or more items, each a JSON object
+  representing an individual attestation.
+* Each attestation object **MUST** be verifiable by the index. If the index fails
   to verify any attestation in ``attestations``, it **MUST** reject the upload.
+  The format of attestation objects is defined under :ref:`attestation-object`
+  and the process for verifying attestations is defined under
+  :ref:`attestation-verification`.
 
-In addition to the above, the index **SHALL** enforce a consistency
-policy for release attestations via the following:
+Index changes
+-------------
+
+Simple Index
+^^^^^^^^^^^^
 
-* If the first file under a new release is supplied with ``attestations``,
-  then all subsequently uploaded files under the same release **MUST** also
-  have ``attestations``. Conversely, if the first file under a new release
-  does not have any ``attestations``, then all subsequent uploads under the
-  same release **MUST NOT** have ``attestations``.
-* All files under the same release **MUST** have the same set of well-known
-  attestation format keys.
+* When an uploaded file has one or more attestations, the index **MAY** include a
+  ``data-provenance`` attribute on its file link.
 
-The index **MUST** reject any file upload that does not satisfy these
-consistency properties.
+* When the ``data-provenance`` attribute is present, its value **SHALL** be
+  a base64 encoded :ref:`provenance object <provenance-object>`.
 
-Index changes
--------------
+JSON-based Simple API
+^^^^^^^^^^^^^^^^^^^^^
+
+* When an uploaded file has one or more attestations, the index **MAY** include a
+  ``provenance`` object in the ``file`` dictionary for that file.
+* ``provenance``, when present, **MUST** be a :ref:`provenance object <provenance-object>`.
+
+These changes require a version change to the JSON API:
+
+* The ``api-version`` must specify version 1.2 or later.
+
+.. _attestation-object:
+
+Attestation objects
+-------------------
+
+An attestation object is a JSON object with several required keys; applications
+or signers may include additional keys so long as all explicitly
+listed keys are provided. The required layout of an attestation
+object is provided as pseudocode below.
+
+.. code-block:: python
+
+  @dataclass
+  class Attestation:
+      version: Literal[1]
+      """
+      The attestation format's version, which is always 1.
+      """
+
+      verification_material: VerificationMaterial
+      """
+      Cryptographic materials used to verify `message_signature`.
+      """
+
+      message_signature: str
+      """
+      The attestation's signature, as `base64(raw-sig)`, where `raw-sig`
+      is the raw bytes of the signing operation.
+      """
+
+  @dataclass
+  class VerificationMaterial:
+      certificate: str
```

Comment:
Do we want to allow for keys, or keep this focused on identity-based signing?

Reply:
I was going to limit this to identity based signing for the "v1" format, just in terms of reducing the scope of the PEP (so that I don't have to specify PyPI's handling of keys, since that doesn't currently exist 😅).

Reply:
I thought about this some more, and unless there's a strong objection to the current identity cert limitation, IMO it's the most tractable approach for the "v1" object format (in terms of keeping the PEP relatively small + reflecting what the initial PyPI implementation will do).

Reply:
SGTM!